linfa-svm/src/solver_smo.rs

```diff
@@ -825,14 +830,31 @@ impl<'a, F: Float, K: 'a + Permutable<F>> SolverState<'a, F, K> {
         };
 
         // put back the solution
-        let alpha: Vec<F> = (0..self.ntotal())
+        // mutable for succesive regression alpha values calculations
+        let mut alpha: Vec<F> = (0..self.ntotal())
             .map(|i| self.alpha[self.active_set[i]].val())
             .collect();
 
+        if self.ntotal() > self.dataset.len_of(Axis(0)) {
+            for i in 0..self.dataset.len_of(Axis(0)) {
+                let tmp = alpha[i + self.dataset.len_of(Axis(0))];
+                alpha[i] -= tmp;
+            }
+            alpha.truncate(self.dataset.len_of(Axis(0)));
+        }
+
+        // Make unmutable
+        let alpha = alpha;
+
+        // This needs to be after the alpha calculation for regression
+        // otherwise it will precompute the dot product with the wrong
+        // values of alpha in the linear kernel case
+
         // if the kernel is linear, then we can pre-calculate the dot product
         let linear_decision = if self.kernel.inner().is_linear() {
-            let mut tmp = Array1::zeros(self.kernel.inner().dataset.len_of(Axis(1)));
-            for (i, elm) in self.kernel.inner().dataset.outer_iter().enumerate() {
+            let mut tmp = Array1::zeros(self.dataset.len_of(Axis(1)));
+
+            for (i, elm) in self.dataset.outer_iter().enumerate() {
                 tmp.scaled_add(self.target(i) * alpha[i], &elm);
             }
 
```

Comment:
Here, when pre-calculating the dop product in the linear case only the first `n` values of alpha are used but in the case of regression there are `2n` alpha values, so you need to do `alpha[i] - alpha[n+i]` before you pre-calculate the dot product, otherwise the coefficients will be wrong

Reply:
ah I see

Reply:
can you add a comment for the code block that this is a special case of eps regression, because the it's difficult to understand on first glance